### PR TITLE
Scalar functions refactor.

### PR DESCRIPTION
**Pull request by fiedukow**

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ReflectionParametricScalar.java [hunk 1]
```diff
@@ -227,13 +277,20 @@ public static SqlScalarFunction parseDefinition(Class<?> clazz)
             }
         }
 
-        requireNonNull(signature, format("No implementations found for %s", name));
+        Map<Signature, Implementation> exactImplementationsMap = exactImplementations.build();
+        if (signature == null) {
+            checkArgument(!exactImplementationsMap.isEmpty(), "Implementation of ScalarFunction %s must be parametric or exact implementation.", objectName);
+            checkArgument(exactImplementationsMap.size() == 1, "It is not allowed to use clases without generic signature.");
+            Map.Entry<Signature, Implementation> onlyImplementation = exactImplementationsMap.entrySet().iterator().next();
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ReflectionParametricScalar.java]: `signature = Iterables.getOnlyElement(exactImplementationsMap.keySet())`

Comment (fiedukow) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ReflectionParametricScalar.java]: It's fixed in later commit ;)

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/annotations/ScalarHeader.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    public boolean isHidden()
+    {
+        return hidden;
+    }
+
+    public boolean isDeterministic()
+    {
+        return deterministic;
+    }
+}
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/annotations/ScalarHeader.java]: new line. Configure your `Intellij` to add this automatically.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ParametricScalar.java [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+import static com.facebook.presto.spi.StandardErrorCode.AMBIGUOUS_FUNCTION_IMPLEMENTATION;
+import static com.facebook.presto.spi.StandardErrorCode.FUNCTION_IMPLEMENTATION_MISSING;
+import static com.facebook.presto.util.Failures.checkCondition;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class ParametricScalar
+        extends SqlScalarFunction
+{
+    ScalarHeader details;
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ParametricScalar.java]: `private final`

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ParametricScalar.java [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    ScalarImplementations implementations;
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ParametricScalar.java]: ditto

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/annotations/ScalarImplementations.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    // These are implementations for concrete types (they have no unbound type parameters), and have the highest priority when picking an implementation
+    private final Map<Signature, ScalarImplementation> exactImplementations;
+    // These are implementations for when a type parameter binds to a specific native container type
+    private final List<ScalarImplementation> specializedImplementations;
+    // These are generic implementations
+    private final List<ScalarImplementation> genericImplementations;
+
+    public ScalarImplementations(Map<Signature, ScalarImplementation> exactImplementations, List<ScalarImplementation> specializedImplementations, List<ScalarImplementation> genericImplementations)
+    {
+        this.exactImplementations = exactImplementations;
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/annotations/ScalarImplementations.java]: As it is public class maybe we should check all the fields for `null`?

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/annotations/ScalarImplementation.java [hunk 6]
```diff
@@ -84,189 +77,190 @@
...
+        this.dependencies = ImmutableList.copyOf(requireNonNull(dependencies, "dependencies is null"));
+        this.constructor = requireNonNull(constructor, "constructor is null");
+        this.constructorDependencies = ImmutableList.copyOf(requireNonNull(constructorDependencies, "constructorDependencies is null"));
+        this.argumentNativeContainerTypes = ImmutableList.copyOf(requireNonNull(argumentNativeContainerTypes, "argumentNativeContainerTypes is null"));
+        this.specializedTypeParameters = ImmutableMap.copyOf(requireNonNull(specializedTypeParameters, "specializedTypeParameters is null"));
     }
 
-    @Override
-    public boolean isHidden()
+    public MethodHandleAndConstructor specialize(Signature boundSignature, BoundVariables boundVariables, TypeManager typeManager, FunctionRegistry functionRegistry)
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/annotations/ScalarImplementation.java]: Make it `Optional`

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/annotations/ScalarHeaderAndMethods.java [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+import static com.google.common.base.Preconditions.checkArgument;
+
+public class ScalarHeaderAndMethods
+{
+    private final ScalarHeader header;
+    private final List<Method> methods;
+
+    public ScalarHeaderAndMethods(ScalarHeader header, List<Method> methods)
+    {
+        this.header = header;
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/annotations/ScalarHeaderAndMethods.java]: check for nulls

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/annotations/ScalarHeaderAndMethods.java [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+        this.methods = methods;
+    }
+
+    public static List<ScalarHeaderAndMethods> fromFunctionDefinitionClassAnnotations(Class annotated)
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/annotations/ScalarHeaderAndMethods.java]: I would rather move those methods to parser, made them private, and leave this class as plain POJO. It is easier to follow parser algorithm if all the methods are close to each other.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/annotations/ScalarHeader.java [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+        checkArgument(false, "Only Classes and Methods are supported as annotated elements.");
+        return null;
+    }
+
+    private static String camelToSnake(String name)
+    {
+        return LOWER_CAMEL.to(LOWER_UNDERSCORE, name);
+    }
+
+    public static List<ScalarHeader> fromAnnotatedElement(AnnotatedElement annotated)
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/annotations/ScalarHeader.java:63]: Move this method to parser as well. Make this object plain POJO, with only data fields and the getters.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/annotations/ScalarHeaderAndMethods.java [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+import com.google.common.collect.ImmutableList;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Method;
+import java.lang.reflect.Modifier;
+import java.util.List;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+public class ScalarHeaderAndMethods
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/annotations/ScalarHeaderAndMethods.java]: Once you move static methods to parser, this class can be moved to the parser to as a internal private class

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/annotations/ScalarHeader.java (hunk 2 above)
Comment (fiedukow) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/annotations/ScalarHeader.java]: done

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ParametricScalar.java (hunk 3 above)
Comment (fiedukow) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ParametricScalar.java]: done

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/ParametricScalar.java (hunk 4 above)
Comment (fiedukow) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/ParametricScalar.java]: done

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/annotations/ScalarImplementations.java (hunk 5 above)
Comment (fiedukow) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/annotations/ScalarImplementations.java]: done

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/annotations/ScalarImplementation.java (hunk 6 above)
Comment (fiedukow) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/annotations/ScalarImplementation.java]: done

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/annotations/ScalarHeaderAndMethods.java (hunk 7 above)
Comment (fiedukow) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/annotations/ScalarHeaderAndMethods.java]: done

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/annotations/ScalarHeader.java (hunk 9 above)
Comment (fiedukow) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/annotations/ScalarHeader.java:63]: Done. I also made ScalarHeaderAndMethod an inner class of Parser as this pair does not make sense in any other context.

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/annotations/ScalarHeaderAndMethods.java (hunk 8 above)
Comment (fiedukow) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/annotations/ScalarHeaderAndMethods.java]: done (comment below)

File: presto-main/src/main/java/com/facebook/presto/operator/scalar/annotations/ScalarHeaderAndMethods.java (hunk 10 above)
Comment (fiedukow) [inline presto-main/src/main/java/com/facebook/presto/operator/scalar/annotations/ScalarHeaderAndMethods.java]: Hah :D I've made it already after previous comment :D :D :D

File: presto-main/src/main/java/com/facebook/presto/metadata/FunctionListBuilder.java [hunk 11]
```diff
@@ -33,7 +33,6 @@
 import com.facebook.presto.type.Constraint;
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/metadata/FunctionListBuilder.java]: Not meaningful commit message.

**Pull request closed without merging**